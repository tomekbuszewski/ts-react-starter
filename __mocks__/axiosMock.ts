import axios from "axios";

export const defaultAxiosAnswer = {
  statusText: "",
  config: {},
  options: {},
  headers: {},
};

export const mockedAxios = axios as jest.Mocked<typeof axios>;
