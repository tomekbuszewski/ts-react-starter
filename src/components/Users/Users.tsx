import * as React from "react";
import { connect } from "react-redux";

import { getUsers } from "@redux/users/actions";
import { IReduxUser } from "@redux/users/reducer";
import { IReduxStore } from "@redux/reducers";

interface Props {
  get: () => void;
  state: number | null;
  users: IReduxUser[];
}

export const UsersComponent = (props: Props) => {
  React.useEffect(() => {
    if (props.state === null) {
      props.get();
    }
  }, []);

  if (props.state === null) {
    return <div data-testid="state-loading">Loading users...</div>;
  }

  if (props.state !== 200) {
    return <div data-testid="state-error">Error</div>;
  }

  return (
    <React.Fragment>
      <h2>Users:</h2>
      <ul data-testid="state-success">
        {props.users.map((user: IReduxUser) => (
          <li key={user.id}>{user.name}</li>
        ))}
      </ul>
    </React.Fragment>
  );
};

const mapState = (state: IReduxStore) => state.users;

const mapDispatch = {
  get: getUsers,
};

export const Users = connect(mapState, mapDispatch)(UsersComponent);
