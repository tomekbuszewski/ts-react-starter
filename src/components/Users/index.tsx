import * as React from "react";
import { connect } from "react-redux";

import { getUsers } from "@redux/users/actions";
import { IReduxUser } from "@redux/users/reducer";
import { IReduxStore } from "@redux/reducers";

interface Props {
  get: () => void;
  state: number | null;
  users: IReduxUser[];
}

class Users extends React.Component<Props> {
  public componentDidMount() {
    if (this.props.state === null) {
      this.props.get();
    }
  }

  public render() {
    if (this.props.state === null) {
      return <div data-testid="state-loading">Loading users...</div>;
    }

    if (this.props.state !== 200) {
      return <div data-testid="state-error">Error</div>;
    }

    return (
      <React.Fragment>
        <h2>Users:</h2>
        <ul data-testid="state-success">
          {this.props.users.map((user: IReduxUser) => (
            <li key={user.id}>{user.name}</li>
          ))}
        </ul>
      </React.Fragment>
    );
  }
}

const mapState = (state: IReduxStore) => ({
  ...state.users,
});

const mapDispatch = {
  get: getUsers,
};

export default connect(
  mapState,
  mapDispatch,
)(Users);
