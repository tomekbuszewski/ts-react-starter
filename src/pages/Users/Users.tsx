import * as React from "react";
import ky from "ky-universal";
import { View } from "./Users.view";

const Users: React.FunctionComponent = () => {
  const [users, setUsers] = React.useState([]);

  React.useEffect(() => {
    (async () => {
      const data = await ky("https://test-endpoint.com/users");
      setUsers(await data.json());
    })();
  }, []);

  return <View users={users} />;
};

export { Users };
