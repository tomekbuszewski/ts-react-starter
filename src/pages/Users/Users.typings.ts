export enum PAGE_USERS_TESTING_IDS {
  USERS_LIST = "page-users-users-list",
  NO_USERS = "page-users-no-users",
}
