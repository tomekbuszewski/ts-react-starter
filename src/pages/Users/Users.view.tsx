import * as React from "react";
import { PAGE_USERS_TESTING_IDS, ViewProps } from "@pages/Users/Users.typings";

const View: React.FunctionComponent<ViewProps> = ({ users }) => {
  if (users.length > 0) {
    return (
      <ul data-testid={PAGE_USERS_TESTING_IDS.USERS_LIST}>
        {users.map((user) => (
          <li key={user.id}>{user.name}</li>
        ))}
      </ul>
    );
  }

  return (
    <div data-testid={PAGE_USERS_TESTING_IDS.NO_USERS}>No users available</div>
  );
};

export { View };
