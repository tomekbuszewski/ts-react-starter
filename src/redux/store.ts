/* eslint-disable no-undef */
import { createStore } from "redux";

import reducers from "./reducers";

export const makeStore = () => {
  const enhancer = (window as any).__REDUX_DEVTOOLS_EXTENSION__ &&
    (window as any).__REDUX_DEVTOOLS_EXTENSION__();

  return createStore(
    reducers,
    enhancer,
  );
};

export default makeStore();
