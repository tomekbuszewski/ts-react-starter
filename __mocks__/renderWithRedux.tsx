import * as React from "react";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";

import { Store } from "redux";

export function renderWithRedux(content: React.ReactNode, store: Store) {
  return {
    ...render(<Provider store={store}>{content}</Provider>),
    store,
  };
}
