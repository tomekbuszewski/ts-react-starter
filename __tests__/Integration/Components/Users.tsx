import * as React from "react";
import { Store } from "redux";
import { render, waitForElement } from "react-testing-library";
import { Provider } from "react-redux";

import { makeStore } from "@redux/store";
import Users from "@components/Users";

import {
  mockedAxios,
  defaultAxiosAnswer as defaultAnswer,
} from "__mocks__/axiosMock";
import { renderWithRedux } from "__mocks__/renderWithRedux";

let store: Store;

const successMockData = {
  ...defaultAnswer,
  status: 200,
  data: [
    {
      id: 1,
      email: "one@one.com",
      username: "one",
      name: "One One",
    },
    {
      id: 2,
      email: "two@two.com",
      username: "two",
      name: "Two Two",
    },
  ],
};

const failMockData = {
  ...defaultAnswer,
  status: 500,
  data: {},
};

jest.mock("axios");

describe("Integration `UsersComponent` tests", () => {
  beforeEach(() => {
    store = makeStore();
  });

  it("should have empty state on mount", () => {
    const component = renderWithRedux(<Users />, store);

    expect(component.container.innerHTML).toBe(
      `<div data-testid="state-loading">Loading users...</div>`,
    );
    expect(component).toMatchSnapshot();
  });

  it("should display users on success", async () => {
    mockedAxios.get.mockResolvedValue({ ...successMockData });

    const component = renderWithRedux(<Users />, store);
    const mountedNode = await waitForElement(() =>
      component.getByTestId("state-success"),
    );

    expect(mountedNode.innerHTML).toBe("<li>One One</li><li>Two Two</li>");
    // expect(component).toMatchSnapshot();
  });

  it("should display error message on failure", async () => {
    mockedAxios.get.mockResolvedValueOnce({ ...failMockData });

    const component = renderWithRedux(<Users />, store);
    const mountedNode = await waitForElement(() =>
      component.getByTestId("state-error"),
    );

    expect(mountedNode.innerHTML).toBe("Error");
    // expect(component).toMatchSnapshot();
  });
});
