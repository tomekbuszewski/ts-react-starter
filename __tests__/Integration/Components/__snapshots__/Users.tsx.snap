// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration \`UsersComponent\` tests should display error message on failure 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        data-testid="state-error"
      >
        Error
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      data-testid="state-error"
    >
      Error
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllBySelectText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getBySelectText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllBySelectText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryBySelectText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Integration \`UsersComponent\` tests should display users on success 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h2>
        Users:
      </h2>
      <ul
        data-testid="state-success"
      >
        <li>
          One One
        </li>
        <li>
          Two Two
        </li>
      </ul>
    </div>
  </body>,
  "container": <div>
    <h2>
      Users:
    </h2>
    <ul
      data-testid="state-success"
    >
      <li>
        One One
      </li>
      <li>
        Two Two
      </li>
    </ul>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllBySelectText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getBySelectText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllBySelectText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryBySelectText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Integration \`UsersComponent\` tests should display users on success 2`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        data-testid="state-loading"
      >
        Loading users...
      </div>
    </div>
    <div>
      <div
        data-testid="state-error"
      >
        Error
      </div>
    </div>
    <div>
      <h2>
        Users:
      </h2>
      <ul
        data-testid="state-success"
      >
        <li>
          One One
        </li>
        <li>
          Two Two
        </li>
      </ul>
    </div>
  </body>,
  "container": <div>
    <h2>
      Users:
    </h2>
    <ul
      data-testid="state-success"
    >
      <li>
        One One
      </li>
      <li>
        Two Two
      </li>
    </ul>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllBySelectText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getBySelectText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllBySelectText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryBySelectText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Integration \`UsersComponent\` tests should have empty state on mount 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        data-testid="state-loading"
      >
        Loading users...
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      data-testid="state-loading"
    >
      Loading users...
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllBySelectText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getBySelectText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllBySelectText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryBySelectText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
